tsdb/exemplar_test.go

```diff
@@ -0,0 +1,538 @@
+// limitations under the License.
+
+package tsdb
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+func TestAddExemplar(t *testing.T) {
+	es, err := NewCircularExemplarStorage(2, nil)
+	require.NoError(t, err)
+
+	l := labels.Labels{
+		{Name: "service", Value: "asdf"},
+	}
+	e := exemplar.Exemplar{
+		Labels: labels.Labels{
+			labels.Label{
+				Name:  "traceID",
+				Value: "qwerty",
+			},
+		},
+		Value: 0.1,
+		Ts:    1,
+	}
+
+	err = es.AddExemplar(l, e)
+	require.NoError(t, err)
+	require.Equal(t, es.index[l.String()].last, 0, "exemplar was not stored correctly")
+
+	e2 := exemplar.Exemplar{
+		Labels: labels.Labels{
+			labels.Label{
+				Name:  "traceID",
+				Value: "zxcvb",
+			},
+		},
+		Value: 0.1,
+		Ts:    2,
+	}
+
+	err = es.AddExemplar(l, e2)
+	require.NoError(t, err)
+	require.Equal(t, es.index[l.String()].last, 1, "exemplar was not stored correctly, location of newest exemplar for series in index did not update")
+	require.True(t, es.exemplars[es.index[l.String()].last].exemplar.Equals(e2), "exemplar was not stored correctly, expected %+v got: %+v", e2, es.exemplars[es.index[l.String()].last].exemplar)
+}
+
+func TestAddDuplicateExemplar(t *testing.T) {
+	es, err := NewCircularExemplarStorage(5, nil)
+	require.NoError(t, err)
+
+	l := labels.Labels{
+		{Name: "service", Value: "asdf"},
+	}
+	e := exemplar.Exemplar{
+		Labels: labels.Labels{
+			labels.Label{
+				Name:  "traceID",
+				Value: "qwerty",
+			},
+		},
+		Value: 0.1,
+		HasTs: true,
+		Ts:    101,
+	}
+
+	es.AddExemplar(l, e)
+	require.True(t, reflect.DeepEqual(es.exemplars[0].exemplar, e), "exemplar was not stored correctly")
+
+	err = es.AddExemplar(l, e)
+	require.Error(t, err, "no error when attempting to add duplicate exemplar")
+	require.True(t, err == storage.ErrDuplicateExemplar, "duplicate exemplar was added")
+}
+
+func TestAddOutOfOrderExemplar(t *testing.T) {
+	es, err := NewCircularExemplarStorage(5, nil)
+	require.NoError(t, err)
+
+	l := labels.Labels{
+		{Name: "service", Value: "asdf"},
+	}
+	e := exemplar.Exemplar{
+		Labels: labels.Labels{
+			labels.Label{
+				Name:  "traceID",
+				Value: "qwerty",
+			},
+		},
+		Value: 0.1,
+		HasTs: true,
+		Ts:    101,
+	}
+	e2 := exemplar.Exemplar{
+		Labels: labels.Labels{
+			labels.Label{
+				Name:  "traceID",
+				Value: "qwerty",
+			},
+		},
+		Value: 0.2,
+		HasTs: true,
+		Ts:    101,
+	}
+
+	es.AddExemplar(l, e)
+	require.True(t, reflect.DeepEqual(es.exemplars[0].exemplar, e), "exemplar was not stored correctly")
+
+	err = es.AddExemplar(l, e2)
+	require.Error(t, err, "no error when attempting to add out of order exemplar")
+	require.True(t, err == storage.ErrOutOfOrderExemplar, "out of order exemplar was added")
+}
+
+func TestAddExtraExemplar(t *testing.T) {
```

Comment:
Did you mean something like `TestQueueOverflow/Overwrite`? Also do we want to test overflow with >1 series?